lib/FileService.php

```diff
@@ -0,0 +1,156 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Richdocuments;
+
+use OC\User\Session;
+use OCP\Files\IRootFolder;
+use OCP\Files\File;
+use OCP\ILogger;
+use OCP\IUserManager;
+use OCA\Richdocuments\AppConfig;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+class FileService {
+	
+	/**
+	 * @var ILogger
+	 */
+	private $logger;
+	
+	/**
+	 * @var AppConfig
+	 */
+	private $appConfig;
+
+	/**
+	 * @var IUserManager
+	 */
+	private $userManager;
+
+	/**
+	 * @var Session
+	 */
+	private $userSession;
+
+	/**
+	 * @var EventDispatcher
+	 */
+	private $eventDispatcher;
+
+	/**
+	 * @var IRootFolder
+	 */
+	private $rootFolder;
+	
+	public function __construct(
+		ILogger $logger,
+		AppConfig $appConfig,
+		IUserManager $userManager,
+		Session $userSession,
+		EventDispatcher $eventDispatcher,
+		IRootFolder $rootFolder
+	) {
+		$this->appConfig = $appConfig;
+		$this->logger = $logger;
+		$this->userManager = $userManager;
+		$this->userSession = $userSession;
+		$this->eventDispatcher = $eventDispatcher;
+		$this->rootFolder = $rootFolder;
+	}
+
+	/**
+	 * Get privileged access to original file handle as user
+	 * for given fileId
+	 *
+	 * @param int $fileId original file id
+	 * @param string $ownerUID original file owner
+	 * @param string $userUID file editor
+	 *
+	 * @return null|File
+	 */
+	public function getFileHandle(int $fileId, string $ownerUID, string $userUID): ?File {
+		if (!$ownerUID) {
+			$this->logger->warning('getFileHandle(): owner must be provided', ['app' => 'richdocuments']);
+			return null;
+		}
+
+		if ($userUID) {
+			$user = $this->userManager->get($userUID);
+			if (!$user) {
+				$this->logger->warning('getFileHandle(): No such user', ['app' => 'richdocuments']);
+				return null;
+			}
+
+			// Make sure editor session is opened for registering activity over file handle
+			$this->logger->debug('getFileHandle(): Register session as ' . $userUID, ['app' => 'richdocuments']);
+			if (!$this->appConfig->encryptionEnabled()) {
+				// Set session for a user
+				$this->userSession->setUser($user);
+			} elseif ($this->appConfig->masterEncryptionEnabled()) {
+				// With master encryption, decryption is based on master key (no user password needed)
+				// make sure audit/activity is triggered for editor session
+				$this->userSession->setUser($user);
+
+				// emit login event to allow decryption of files via master key
+				$afterEvent = new GenericEvent(null, ['loginType' => 'password', 'user' => $user, 'uid' => $userUID, 'password' => '']);
```

Comment:
I guess we'll need to think about this at some point because a login event shouldn't be triggered here.
Probably no short term action to do for now, but we shouldn't figure out something to do this properly.

Reply:
Yes, we would need to do this in Core. I think this was discussed longer time ago, but I think this would need a rework of the encryption stuff in core to do this correctly, which is a bit risky.

Reply:
adjusted